src/sentry/hybridcloud/tasks/deliver_webhooks.py

```diff
@@ -0,0 +1,275 @@
+from rest_framework import status
+
+from sentry.exceptions import RestrictedIPAddress
+from sentry.hybridcloud.models.webhookpayload import BACKOFF_INTERVAL, MAX_ATTEMPTS, WebhookPayload
+from sentry.shared_integrations.exceptions import (
+    ApiConflictError,
+    ApiConnectionResetError,
+    ApiError,
+    ApiHostError,
+    ApiTimeoutError,
+)
+from sentry.silo.base import SiloMode
+from sentry.silo.client import RegionSiloClient, SiloClientError
+from sentry.tasks.base import instrumented_task
+from sentry.types.region import get_region_by_name
+from sentry.utils import json, metrics
+
+logger = logging.getLogger(__name__)
+
+MAX_MAILBOX_DRAIN = 50
+"""
+The maximum number of records that will be delivered in a scheduled delivery
+
+There is a balance here between clearing big backlogs and having races when
+a batch is slow but not timeout slow.
+"""
+
+BATCH_SCHEDULE_OFFSET = datetime.timedelta(minutes=BACKOFF_INTERVAL)
+"""
+The time that batches are scheduled into the future when work starts.
+Spacing batches out helps minimize competitive races when delivery is slow
+but not at the timeout threshold
+"""
+
+BATCH_SIZE = 1000
+"""The number of mailboxes that will have messages scheduled each cycle"""
+
+
+class DeliveryFailed(Exception):
+    """
+    Used to signal an expected delivery failure.
+    """
+
+    pass
+
+
+@instrumented_task(
+    name="sentry.hybridcloud.tasks.deliver_webhooks.schedule_webhook_delivery",
+    queue="webhook.control",
+    silo_mode=SiloMode.CONTROL,
+)
+def schedule_webhook_delivery(**kwargs) -> None:
+    """
+    Find mailboxes that contain undelivered webhooks that were scheduled
+    to be delivered now or in the past.
+
+    Triggered frequently by celery beat.
+    """
+    # The double call to .values() ensures that the group by includes mailbox_nam
+    # but only id_min is selected
+    head_of_line = (
+        WebhookPayload.objects.all()
+        .values("mailbox_name")
+        .annotate(id_min=Min("id"))
+        .values("id_min")
+    )
+    # Get any heads that are scheduled to run
+    scheduled_mailboxes = WebhookPayload.objects.filter(
+        schedule_for__lte=timezone.now(),
+        id__in=Subquery(head_of_line),
+    ).values("id", "mailbox_name")
+
+    metrics.distribution(
+        "hybridcloud.schedule_webhook_delivery.mailbox_count", scheduled_mailboxes.count()
+    )
+    for record in scheduled_mailboxes[:BATCH_SIZE]:
+        # Reschedule the records that we will attempt to deliver next.
+        # We reschedule in an attempt to minimize races for potentially in-flight batches.
+        mailbox_batch = (
+            WebhookPayload.objects.filter(id__gte=record["id"], mailbox_name=record["mailbox_name"])
+            .order_by("id")
+            .values("id")[:MAX_MAILBOX_DRAIN]
+        )
+        WebhookPayload.objects.filter(id__in=Subquery(mailbox_batch)).update(
+            schedule_for=timezone.now() + BATCH_SCHEDULE_OFFSET
+        )
```

Comment:
Because we're delaying tasks to process each batch, I wanted to offset the next schedule time for batches so that subsequent scheduling jobs don't re-queue the same work. The assumption is that each batch should complete in 3 minutes generally. If a service is very slow but not timeout slow we could end up with competitive races.

A potential solution to this could be to have deliver tasks maintain a deadline that they need to complete their work in. Should they have unfinished work by the deadline the task would abort early. I'm not sure the complexity is justified just yet though.